src/neurostatslib/glm.py

```diff
@@ -1,431 +1,598 @@
+        # (n_samples, n_neurons)
+        feed_forward_contrib = jnp.einsum("ik,tik->ti", Wf, feedforward_input)
+
+        def scan_fn(
+            data: Tuple[jnp.ndarray, int], key: jax.random.PRNGKeyArray
+        ) -> Tuple[Tuple[jnp.ndarray, int], Tuple[jnp.ndarray, jnp.ndarray]]:
+            """Scan over time steps and simulate activity and rates.
+
+            This function simulates the neural activity and firing rates for each time step
+            based on the previous activity, feedforward input, and model coefficients.
+            """
+            activity, chunk = data
+
+            # Convolve the neural activity with the coupling basis matrix
+            conv_act = convolve_1d_trials(coupling_basis_matrix, activity[None, :, :])[
+                0
+            ]
+
+            # Extract the corresponding slice of the feedforward input for the current time step
+            input_slice = jax.lax.dynamic_slice(
+                feed_forward_contrib,
+                (chunk, 0),
+                (1, feed_forward_contrib.shape[1]),
             )
 
-        subkeys = jax.random.split(random_key, num=n_timesteps)
+            # Reshape the convolved activity and concatenate with the input slice to form the model input
+            conv_act = jnp.tile(
+                conv_act.reshape(conv_act.shape[0], -1), conv_act.shape[1]
+            ).reshape(conv_act.shape[0], conv_act.shape[1], -1)
 
-        def scan_fn(spikes, key):
-            # (n_neurons, n_basis_funcs, 1)
-            X = jnp.transpose(
-                convolve_1d_trials(self.spike_basis_matrix.T, spikes.T[None, :, :])[0],
-                (1, 2, 0),
+            # Predict the firing rate using the model coefficients
+            # Doesn't use predict because the non-linearity needs
+            # to be applied after we add the feed forward input
+            firing_rate = self.noise_model.inverse_link_function(
+                jnp.einsum("ik,tik->ti", Wr, conv_act) + input_slice + bs[None, :]
             )
-            fr = self._predict((Ws, bs), X).squeeze(-1)
-            new_spikes = jax.random.poisson(key, fr)
-            concat_spikes = jnp.column_stack((spikes[:, 1:], new_spikes))
-            return concat_spikes, new_spikes
 
-        _, simulated_spikes = jax.lax.scan(scan_fn, init_spikes, subkeys)
+            # Simulate activity based on the predicted firing rate
+            new_act = self.noise_model.emission_probability(key, firing_rate)
 
-        return simulated_spikes.T
+            # Prepare the spikes for the next iteration (keeping the most recent spikes)
+            concat_act = jnp.row_stack((activity[1:], new_act)), chunk + 1
+            return concat_act, (new_act, firing_rate)
+
+        _, outputs = jax.lax.scan(scan_fn, (init_y, 0), subkeys)
+        simulated_activity, firing_rates = outputs
+        return jnp.squeeze(simulated_activity, axis=1), jnp.squeeze(
```

Comment:
why is squeeze necessary here? (what's the dummy dimension being used for earlier?

Reply:
because at each iteration of scan the simulate counts and firing rate will have shape (1, n_neurons), where 1 is the time point that you are simulating. Scan will concatenate all samples in an array (n_samples, 1, n_neurons).

The other way of doing this is squeezing insite the scan function so that the rate and counts have dimension (n_neurons, )

Reply:
In the process of doing this, I realize that duplicating the convolved spikes could be avoided; this ended up simplifying the algorithm too, and probably made it more efficient, since it doesn't require any writing on memory of copy of the same stuff